File: cpp/source/communication/DebugData.cpp
#include "DebugData.hpp"

#include <string.h>

#ifdef __SKYDIVE_USE_STL__

#include <iomanip>

#include "SkyException.hpp"

#endif //__SKYDIVE_USE_STL__

DebugData::DebugData(void)
{
}

DebugData::DebugData(const unsigned char* src)
{
    memcpy((unsigned char*)this + 4, src, getDataSize());
}

IMessage::PreambleType DebugData::getPreambleType(void) const
{
    return IMessage::CONTROL;
}

void DebugData::serialize(unsigned char* dst) const
{
    memcpy(dst, (unsigned char*)this + 4, getDataSize());
}

IMessage::MessageType DebugData::getMessageType(void) const
{
    return DEBUG_DATA;
}

unsigned DebugData::getDataSize(void) const
{
    return sizeof(DebugData) - 4;
}

void DebugData::setEuler(const Vect3Df& _euler)
{
    euler = _euler;
}

void DebugData::setPosition(const Vect2Df& _position)
{
    position = _position;
}

void DebugData::setAltitude(const float _relativeAltitude, const float _absoluteAltitude)
{
    relativeAltitude = _relativeAltitude;
    absoluteAltitude = _absoluteAltitude;
}

void DebugData::setVerticalVelocity(const float velocity)
{
    verticalVelocity = velocity;
}

void DebugData::setVelocity(const float _velocity)
{
    velocity = _velocity;
}

void DebugData::setUsedThrottle(const float _usedThrottle)
{
    usedThrottle = _usedThrottle;
}

void DebugData::setDistanceToBase(const float _distanceToBase)
{
    distanceToBase = _distanceToBase;
}

void DebugData::setControllerState(const ControllerState& _controllerState)
{
    controllerState = static_cast<unsigned short>(_controllerState);
}

void DebugData::setSolverMode(const ControlData::SolverMode& solverMode)
{
    // clean previous state
    flagsObj.getFlagsVector() &= 0xFC;
    // set new solver flags
    flagsObj.getFlagsVector() |= solverMode;
}

void DebugData::setBatteryVoltage(const float voltage)
{
    const float maxVoltage = 3.3f * 11.0f;
    battery = static_cast<unsigned char>(((255 * voltage) / maxVoltage) + 0.5f);
}

const Vect3Df& DebugData::getEuler(void) const
{
    return euler;
}

const Vect2Df& DebugData::getPosition(void) const
{
    return position;
}

float DebugData::getRelativeAltitude(void) const
{
    return relativeAltitude;
}

float DebugData::getAbsoluteAltitude(void) const
{
    return absoluteAltitude;
}

float DebugData::getVerticalVelocity(void) const
{
    return verticalVelocity;
}

float DebugData::getVelocity(void) const
{
    return velocity;
}

float DebugData::getUsedThrottle(void) const
{
    return usedThrottle;
}

float DebugData::getDistanceToBase(void) const
{
    return distanceToBase;
}

DebugData::ControllerState DebugData::getControllerState(void) const
{
    return static_cast<ControllerState>(controllerState);
}

ControlData::SolverMode DebugData::getSolverMode(void) const
{
    return static_cast<ControlData::SolverMode>(flagsObj.getFlagsVector() & 0x03);
}

float DebugData::getBatteryVoltage(void) const
{
    const float maxVoltage = 3.3f * 11.0f;
    return (battery * maxVoltage) / 255.0f;
}

const Flags<unsigned char>& DebugData::flags(void) const
{
    return flagsObj;
}

Flags<unsigned char>& DebugData::flags(void)
{
    return flagsObj;
}

void DebugData::setGpsFlags(const GpsFix& gpsFix)
{
    switch (gpsFix)
    {
    case NO_FIX:
        flagsObj.setFlagState(GPS_FIX, false); // no GPS fix
        flagsObj.setFlagState(GPS_FIX_3D, false); // no 3D fix
        break;
    case FIX:
        flagsObj.setFlagState(GPS_FIX, true); // GPS fixed
        flagsObj.setFlagState(GPS_FIX_3D, false); // no 3D fix
        break;
    case FIX_3D:
        flagsObj.setFlagState(GPS_FIX, true); // GPS fixed
        flagsObj.setFlagState(GPS_FIX_3D, true); // 3D fixed
        break;
    }
}

void DebugData::setConnectionLost()
{
    controllerState = ControlData::ERROR_CONNECTION;
}

bool DebugData::isGpsFixed(void) const
{
    return flagsObj.getFlagState(GPS_FIX);
}

bool DebugData::isGps3DFixed(void) const
{
    return flagsObj.getFlagState(GPS_FIX_3D);
}

bool DebugData::isBatteryWaring(void) const
{
    return flagsObj.getFlagState(LOW_BATTERY_VOLTAGE);
}

bool DebugData::isAutopilotUsed(void) const
{
    return flagsObj.getFlagState(AUTOPILOT_ENABLED);
}

bool DebugData::isAutolanding(void) const
{
    return flagsObj.getFlagState(AUTOLANDING_ENABLED);
}

bool DebugData::isErrorHandling(void) const
{
    return flagsObj.getFlagState(ERROR_HANDLING);
}

bool DebugData::isDrawable(void) const
{
    // euler angles
    if (euler.x < -roboLib::pi || euler.x > roboLib::pi)
    {
        return false;
    }
    if (euler.y < -roboLib::pi / 2 || euler.y > roboLib::pi / 2)
    {
        return false;
    }
    // position
    if (position.x < -90.0f || position.x > 90.0f)
    {
        return false;
    }
    if (position.y < -90.0f || position.y > 90.0f)
    {
        return false;
    }
    return true;
}

bool DebugData::isConnection(void) const
{
    return controllerState != ControlData::ERROR_CONNECTION;
}

float DebugData::getNormalYaw(void) const
{
    float yaw = getEuler().z;
    if (yaw > 2 * roboLib::pi) yaw -= float(2 * roboLib::pi);
    else if (yaw < 0.0f) yaw += float(2 * roboLib::pi);
    return yaw;
}

#ifdef __SKYDIVE_USE_STL__

std::string DebugData::getControllerStateString(const ControllerState& state)
{
    switch (state)
    {
    case IDLE:
        return std::string("Idle");
    case MANUAL:
        return std::string("Manual");
    case AUTOLANDING:
        return std::string("Autolanding");
    case AUTOLANDING_AP:
        return std::string("Autolanding AP");
    case HOLD_ALTITUDE:
        return std::string("Hold: altitude");
    case HOLD_POSITION:
        return std::string("Hold: position");
    case BACK_TO_BASE:
        return std::string("Back to base");
    case VIA_ROUTE:
        return std::string("Via route");
    case STOP:
        return std::string("STOP");
    case APPLICATION_LOOP:
        return std::string("App. loop");
    default:
        return std::string("ERROR");
    }
}

DebugData DebugData::parseFromString(const std::string& line)
{
    DebugData debugData;
    std::istringstream iss(line);
    float time;
    int battery;
    if (!(iss
          >> time
          >> debugData.euler.x >> debugData.euler.y >> debugData.euler.z
          >> debugData.position.x >> debugData.position.y
          >> debugData.relativeAltitude >> debugData.absoluteAltitude
          >> debugData.velocity
          >> debugData.controllerState >> debugData.flagsObj.getFlagsVector()
          >> battery
          ))
    {
        __SKY_EXCEPTION__("Error while parsing data.");
    }
    debugData.battery = (unsigned char)battery;
    return debugData;
}

std::ostream& operator << (std::ostream& stream, const DebugData& debugData)
{
    stream << std::setprecision(10);
    stream << debugData.getEuler().x << " " << debugData.getEuler().y << " " << debugData.getEuler().z << " ";
    stream << debugData.getPosition().x << " " << debugData.getPosition().y << " ";
    stream << debugData.getRelativeAltitude() << " " << debugData.getAbsoluteAltitude() << " ";
    stream << debugData.getVelocity() << " " << debugData.getControllerState() << " ";
    stream << (int)debugData.flagsObj.getFlagsVector() << " " << debugData.getBatteryVoltage();
    return stream;
}

#endif //__SKYDIVE_USE_STL__

File: java/data/DebugData.java
package com.skydive.java.data;

import com.skydive.java.CommMessage;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Created by Bartosz Nawrot on 2016-08-22.
 * Container tha stores most important telemetry data from drone:
 * - euler angles - rotation of drone
 * - position (lat, lon, alt), altitude is relative to start
 * - vLoc - speed of drone relatively to ground
 * - controller state - actual control command used by controller
 * - flags:
 * GPS fix | GPS 3D fix | low. bat. volt. | errorHandling | autopilotUsed | solver1 | solver2
 * - battery voltage [volts]
 */
public class DebugData {

    private float roll, pitch, yaw;

    private float latitude, longitude;
    private float relativeAltitude; // altitude relative to base location
    private float absoluteAltitude; // altitude above see level
    private float verticalVelocity;

    private float velocity; // velocity relative to XY earth plane [m/s]

    private float usedThrottle;
    private float distanceToBase; // [km]

    private ControllerState controllerState;
    private Flags flags;

    private byte battery;

    public DebugData() {
        this.flags = new Flags(8);
    }

    public DebugData(DebugData debugData) {
        this.roll = debugData.getRoll();
        this.pitch = debugData.getPitch();
        this.yaw = debugData.getYaw();
        this.latitude = debugData.getLatitude();
        this.longitude = debugData.getLongitude();
        this.relativeAltitude = debugData.getRelativeAltitude();
        this.absoluteAltitude = debugData.getAbsoluteAltitude();
        this.verticalVelocity = debugData.getVerticalVelocity();
        this.velocity = debugData.getVelocity();
        this.usedThrottle = debugData.getUsedThrottle();
        this.distanceToBase = debugData.getDistanceToBase();
        this.controllerState = debugData.getControllerState();
        this.flags = new Flags(8, debugData.getFlags());
        this.battery = debugData.getBattery();
    }

    public DebugData(final CommMessage message) {
        ByteBuffer buffer = ByteBuffer.wrap(message.getPayload());
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        this.roll = buffer.getFloat();
        this.pitch = buffer.getFloat();
        this.yaw = buffer.getFloat();
        this.latitude = buffer.getFloat();
        this.longitude = buffer.getFloat();
        this.relativeAltitude = buffer.getFloat();
        this.absoluteAltitude = buffer.getFloat();
        this.verticalVelocity = buffer.getFloat();
        this.velocity = buffer.getFloat();
        this.usedThrottle = buffer.getFloat();
        this.distanceToBase = buffer.getFloat();
        this.controllerState = ControllerState.getControllerState(buffer.getShort());
        this.flags = new Flags(8, buffer.get());
        this.battery = buffer.get();
    }

    public float getRoll() {
        return roll;
    }

    public void setRoll(float roll) {
        this.roll = roll;
    }

    public float getPitch() {
        return pitch;
    }

    public void setPitch(float pitch) {
        this.pitch = pitch;
    }

    public float getYaw() {
        return yaw;
    }

    public void setYaw(float yaw) {
        this.yaw = yaw;
    }

    public float getNormalYaw() {
        float yaw = this.yaw;
        if (yaw > 2 * Math.PI) {
            yaw -= (float) (2 * Math.PI);
        } else if (yaw < 0.0f) {
            yaw += (float) (2 * Math.PI);
        }
        return yaw;
    }

    public float getLatitude() {
        return latitude;
    }

    public void setLatitude(float latitude) {
        this.latitude = latitude;
    }

    public float getLongitude() {
        return longitude;
    }

    public void setLongitude(float longitude) {
        this.longitude = longitude;
    }

    public float getRelativeAltitude() {
        return relativeAltitude;
    }

    public void setRelativeAltitude(float relativeAltitude) {
        this.relativeAltitude = relativeAltitude;
    }

    public float getAbsoluteAltitude() {
        return absoluteAltitude;
    }

    public void setAbsoluteAltitude(float absoluteAltitude) {
        this.absoluteAltitude = absoluteAltitude;
    }

    public float getVerticalVelocity() {
        return verticalVelocity;
    }

    public void setVerticalVelocity(float verticalVelocity) {
        this.verticalVelocity = verticalVelocity;
    }

    public float getVelocity() {
        return velocity;
    }

    public void setVelocity(float velocity) {
        this.velocity = velocity;
    }

    public float getUsedThrottle() {
        return usedThrottle;
    }

    public void setUsedThrottle(float usedThrottle) {
        this.usedThrottle = usedThrottle;
    }

    public float getDistanceToBase() {
        return distanceToBase;
    }

    public void setDistanceToBase(float distanceToBase) {
        this.distanceToBase = distanceToBase;
    }

    public ControllerState getControllerState() {
        return controllerState;
    }

    public void setControllerState(ControllerState controllerState) {
        this.controllerState = controllerState;
    }

    public byte getFlags() {
        return (byte) flags.getFlags();
    }

    public void setFlags(byte flags) {
        this.flags = new Flags(8, flags);
    }

    public boolean getFlagState(FlagId id) {
        return flags.getFlagState(id.getValue());
    }

    public void setFLagState(FlagId id, boolean state) {
        flags.setFlagsState(id.getValue(), state);
    }

    public void setSolverMode(final ControlData.SolverMode solverMode) {
        // clean previous state
        flags.setFlags(flags.getFlags() & 0xFC);
        // set new solver flags
        flags.setFlags(flags.getFlags() | solverMode.getValue());
    }

    public ControlData.SolverMode getSolverMode() {
        return ControlData.SolverMode.getSolverMode((byte) (flags.getFlags() & 0x03));
    }

    public byte getBattery() {
        return battery;
    }

    public void setBattery(byte battery) {
        this.battery = battery;
    }

    public boolean isStopState() {
        return this.controllerState == ControllerState.STOP;
    }

    @Override
    public String toString() {
        return "DebugData:[ roll: " + String.valueOf(roll)
                + ", pitch: " + String.valueOf(pitch)
                + ", yaw: " + String.valueOf(yaw)
                + ", lat: " + String.valueOf(latitude)
                + ", lon: " + String.valueOf(longitude)
                + ", alt: " + String.valueOf(relativeAltitude)
                + ", Controller state: " + controllerState.toString() + "(" + String.valueOf(controllerState.getValue()) + ")"
                + " ]";
    }

    public CommMessage getMessage() {
        ByteBuffer buffer = ByteBuffer.allocate(CommMessage.getPayloadSizeByType(CommMessage.MessageType.CONTROL));
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        buffer.putFloat(getRoll());
        buffer.putFloat(getPitch());
        buffer.putFloat(getYaw());
        buffer.putFloat(getLatitude());
        buffer.putFloat(getLongitude());
        buffer.putFloat(getRelativeAltitude());
        buffer.putFloat(getAbsoluteAltitude());
        buffer.putFloat(getVerticalVelocity());
        buffer.putFloat(getVelocity());
        buffer.putFloat(getUsedThrottle());
        buffer.putFloat(getDistanceToBase());
        buffer.putShort(getControllerState().getValue());
        buffer.put((byte) flags.getFlags());
        buffer.put(battery);
        return new CommMessage(CommMessage.MessageType.CONTROL, buffer.array());
    }

    public enum ControllerState {
        IDLE((short) 0),
        // manual control
        MANUAL(ControlData.ControllerCommand.MANUAL.getValue()),
        // auto lading with specified descend rate
        AUTOLANDING(ControlData.ControllerCommand.AUTOLANDING.getValue()),
        // auto lading with specified descend rate and position hold
        AUTOLANDING_AP(ControlData.ControllerCommand.AUTOLANDING_AP.getValue()),
        // auto altitude hold, throttle value is now specifying descend/climb rate
        // th = 0 -> -v, th = 0.5 -> 0, th = 1.0 -> v
        HOLD_ALTITUDE(ControlData.ControllerCommand.HOLD_ALTITUDE.getValue()),
        // auto position hold, (hold altitude enabled)
        HOLD_POSITION(ControlData.ControllerCommand.HOLD_POSITION.getValue()),
        // autonomous back to base, climb 10 meters above start, cruise to base and auto land with AP
        BACK_TO_BASE(ControlData.ControllerCommand.BACK_TO_BASE.getValue()),
        // cruise via specific route and back to base
        VIA_ROUTE(ControlData.ControllerCommand.VIA_ROUTE.getValue()),
        // immediate STOP (even when fling)
        STOP(ControlData.ControllerCommand.STOP.getValue()),

        APPLICATION_LOOP((short) 3000),

        ERROR_CONNECTION(ControlData.ControllerCommand.ERROR_CONNECTION.getValue());

        private final short value;

        ControllerState(short value) {
            this.value = value;
        }

        public short getValue() {
            return value;
        }

        static public ControllerState getControllerState(short value) {
            for (ControllerState controllerState : ControllerState.values()) {
                if (controllerState.getValue() == value) {
                    return controllerState;
                }
            }
            return IDLE; // TODO throw some exception
        }

        @Override
        public String toString() {
            if (value == IDLE.getValue()) return "Idle";
            else if (value == MANUAL.getValue()) return "Manual";
            else if (value == AUTOLANDING.getValue()) return "Auto landing";
            else if (value == AUTOLANDING_AP.getValue()) return "Auto landing AP";
            else if (value == HOLD_ALTITUDE.getValue()) return "Hold altitude";
            else if (value == HOLD_POSITION.getValue()) return "Hold position";
            else if (value == BACK_TO_BASE.getValue()) return "Back to base";
            else if (value == VIA_ROUTE.getValue()) return "Via route";
            else if (value == STOP.getValue()) return "Stop";
            else if (value == APPLICATION_LOOP.getValue()) return "Application loop";
            else if (value == ERROR_CONNECTION.getValue()) return "Error connection";
            else return "Error type!";
        }
    }

    public enum FlagId {
        SOLVER_1(0),
        SOLVER_2(1),
        AUTOLANDING_ENABLED(2),
        AUTOPILOT_ENABLED(3),
        ERROR_HANDLING(4),
        LOW_BATTERY_VOLTAGE(5),
        GPS_FIX_3D(6),
        GPS_FIX(7);

        private final int value;

        FlagId(int value) {
            this.value = value;
        }

        int getValue() {
            return value;
        }
    }
}
